R/derive_date_vars.R

```diff
@@ -101,214 +117,730 @@
-      mo <- list(first = "01", mid = "06", last = "12")[[date_imputation]]
-    } else {
-      # otherwise, use time_imputation input
-      day__ <- as.integer(sub(".*-", "", date_imputation))
-      mo__ <- as.integer(sub("-.*", "", date_imputation))
-      # check input for day and moth are valid
-      assert_that(is_valid_day(day__))
-      assert_that(is_valid_month(mo__))
-
-      d <- sprintf("%02d", day__)
-      mo <- sprintf("%02d", mo__)
+  if (length(dtc) == 0) {
+    return(vector("character"))
+  }
+
+  # Parse character date ----
+  partial <- get_partialdatetime(dtc)
+  components <- names(partial)
+
+  # Handle preserve argument ----
+  if (!preserve) {
+    for (i in 2:6) {
+      partial[[i]] <- if_else(is.na(partial[[i - 1]]), NA_character_, partial[[i]])
     }
+  }
+  # Determine target components ----
+  target_date <- get_imputation_target_date(
+    date_imputation = date_imputation,
+    month = partial[["month"]]
+  )
+  target_time <- get_imputation_target_time(
+    time_imputation = time_imputation
+  )
+  target <- c(target_date, target_time)
 
-    imputed_date <- case_when(
-      !valid_dtc ~ NA_character_,
-      n_chr >= 10 ~ substr(dtc, 1, 10),
-      n_chr == 9 ~ paste0(substr(dtc, 1, 4), "-", mo, "-", d),
-      n_chr == 7 ~ paste0(dtc, "-", d),
-      n_chr == 4 ~ paste0(dtc, "-", mo, "-", d)
+  for (c in components) {
+    if (highest_imputation < dtm_level(imputation_levels[[c]])) {
+      target[[c]] <- "xx"
+    }
+  }
+
+  # Impute ----
+  imputed <- vector("list", 6)
+  names(imputed) <- components
+  for (c in components) {
+    imputed[[c]] <- if_else(is.na(partial[[c]]), target[[c]], partial[[c]])
+  }
+
+  imputed_dtc <-
+    paste0(
+      paste(imputed[["year"]], imputed[["month"]], imputed[["day"]], sep = "-"),
+      "T",
+      paste(imputed[["hour"]], imputed[["minute"]], imputed[["second"]], sep = ":")
+    )
+
+  imputed_dtc <-
+    if_else(
+      str_detect(imputed_dtc, "x"),
+      NA_character_,
+      imputed_dtc
+    )
+
+  if (date_imputation == "last") {
+    imputed_dtc <-
+      if_else(
+        is.na(partial[["day"]]),
+        strftime(
+          rollback(ymd_hms(imputed_dtc) + months(1)),
+          format = "%Y-%m-%dT%H:%M:%S",
+          tz = "UTC"
+        ),
+        imputed_dtc
+      )
+  }
+
+  # Handle min_dates and max_dates argument ----
+  restrict_imputed_dtc_dtm(
+    dtc,
+    imputed_dtc = imputed_dtc,
+    min_dates = min_dates,
+    max_dates = max_dates
+  )
+}
+
+#' Create a `dtm_level` object
+#'
+#' @param level Datetime level
+#'
+#'   *Permitted Values*: `"Y"` (year, highest level), `"M"` (month), `"D"`
+#'   (day), `"h"` (hour), `"m"` (minute), `"s"` (second, lowest level), `"n"`
+#'   (none)
+#'
+#' @returns A `dtm_level` object
+#'
+#' @details A `dtm_level` object is an ordered factor, i.e., two objects can be
+#'   compared.
+#'
+#' @author Stefan Bundfuss
+dtm_level <- function(level) {
```

Comment:
Should we have this function as `get_dtm_level` just so the other functions are grouped together like `get_partialdatetime`